### PR TITLE
Modify parsing of prefixes

### LINKED ISSUE #307: Modify parsing of prefixes
Currently handling of incorrect prefixes is inconsistent. Some incorrect prefixes, such as `PREFIX_DATE` in `ParserUtil::parseBookmarkTransactionBuilder` are detected and `ParseException` is thrown, while in other cases, known prefixes are simply not detected and treated as part of the value of its preceding prefix.

### PR DESCRIPTION
**Pull request by yongping827**
Resolves #307.

Changes:
* `ArgumentTokenizer::tokenize` now tokenizes all prefixes in `CliSyntax` and throws `ParseException` if an invalid prefix is detected.
* Move `TransactionBuilder` from `test` to `main` package to facilitate `ParseUtil::parseTransactionBuilder`.
* Add `ParseUtil::parseTransactionBuilder` to remove duplicate code in the original `AddExpenseCommandParser` and `AddIncomeCommandParser`. Add additional tests for this.
* Add `ArgumentTokenizerStub` in `ArgumentTokenizerTest` to address the issue of  `ArgumentTokenizerTest` using its own prefixes instead of prefixes in `CliSyntax`.

### CODE REVIEW

**Review by zhaojj2209 — APPROVED** [review]
LGTM!

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/CliSyntax.java [hunk 1]
```diff
@@ -19,4 +19,15 @@
...
+    /**
+     * Retrieves a list containing all prefixes.
+     *
+     * @return List of all prefixes
+     */
+    public static Prefix[] getAllPrefixes() {
+        return new Prefix[] {PREFIX_TITLE, PREFIX_AMOUNT, PREFIX_DATE, PREFIX_CATEGORY, PREFIX_AMOUNT_FROM,
+            PREFIX_AMOUNT_TO, PREFIX_DATE_FROM, PREFIX_DATE_TO};
+    }
+
```
Comment (ianyong) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/CliSyntax.java:32]:
```suggestion
```

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/ArgumentTokenizer.java [hunk 2]
```diff
@@ -17,19 +19,50 @@
...
+            throws ParseException {
+        List<PrefixPosition> positions = findAllPrefixPositions(argsString, prefixFullSet);
+        ArgumentMultimap argumentMultimap = extractArguments(argsString, positions);
+
+        List<Prefix> prefixComplementList = Arrays.stream(prefixFullSet).collect(Collectors.toList());
+        List<Prefix> prefixSubList = Arrays.stream(prefixes).collect(Collectors.toList());
+        prefixComplementList.removeAll(prefixSubList);
+
+        if (argumentMultimap.areAnyPrefixesPresent(prefixComplementList.toArray(new Prefix[0]))) {
+            throw new ParseException("This is an exception message.");
```
Comment (ianyong) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/ArgumentTokenizer.java:63]: Use a more descriptive message.

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/ArgumentTokenizer.java [hunk 3]
```diff
@@ -17,19 +19,50 @@
...
-     * respective argument values. Only the given prefixes will be recognized in the arguments string.
+     * respective argument values. All prefixes from {@code CliSyntax} are recognized, but only the specified
+     * {@param prefixes} are considered valid. The other prefixes not specified are considered invalid and, if detected,
+     * will throw {@code ParseException}
      *
      * @param argsString Arguments string of the form: {@code preamble <prefix>value <prefix>value ...}
-     * @param prefixes   Prefixes to tokenize the arguments string with
+     * @param prefixes   Valid prefixes from the set of prefixes in {@code CliSyntax} to tokenize the arguments with
      * @return           ArgumentMultimap object that maps prefixes to their arguments
+     * @throws           ParseException
```
Comment (ianyong) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/ArgumentTokenizer.java:33]: Add full stops,

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/ArgumentTokenizer.java [hunk 4]
```diff
@@ -17,19 +19,50 @@
...
+     * respective argument values. All prefixes from {@param prefixFullSet} are recognized, but only the specified
+     * {@param prefixes} are considered valid. The other prefixes from {@param prefixFullSet} that are not specified are
+     * considered invalid and, if detected, will throw {@code ParseException}
+     *
+     * @param argsString    Arguments string of the form: {@code preamble <prefix>value <prefix>value ...}
+     * @param prefixFullSet Set of all prefixes to tokenize the arguments with
+     * @param prefixes      Subset of prefixFullSet that are considered valid. Any other prefixes is invalid and
+     *                      thus should throw ParseException
+     * @return              ArgumentMultimap object that maps prefixes to their arguments
+     * @throws              ParseException
```
Comment (ianyong) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/ArgumentTokenizer.java:51]: Add full stops.

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/CliSyntax.java [hunk 5]
```diff
@@ -19,4 +19,15 @@
      * Prevents instantiation of this class.
      */
     private CliSyntax() {}
+
+    /**
+     * Retrieves a list containing all prefixes.
+     *
+     * @return List of all prefixes
```
Comment (ianyong) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/CliSyntax.java:26]:
```suggestion
     * @return List of all prefixes.
```

**Review by ianyong — CHANGES_REQUESTED** [review]

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/CliSyntax.java (hunk 5 above)
Comment (yongping827) [reply src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/CliSyntax.java:26]: Full stop is not required in this case.

Comment (ianyong) [reply src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/CliSyntax.java:26]: Please see https://se-education.org/guides/conventions/java/index.html#comments.

**Review by ianyong — APPROVED** [review]
LGTM!

**Review by zhaojj2209 — APPROVED** [review]
LGTM!

**Pull request merged**